Adds Neon Carpet

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Adds a couple varieties of neon carpet.
- Makes decals care about their plane in addition to their layer.
- Adds a decal subtype that can smooth with the parent.
- Reworks emissive overlays and emissive blockers a bit so they can be applied to turfs, items, and mobs.
  - As a bonus, structures that use vis_overlays shouldn't axe their own emissive blockers!

There is an issue with how emissive decals stack when it comes to the tile items. Specifically the emissive decals on the tiles tend to glow _through_ other items. As far as anyone's been able to tell it's not fixable in DM without convincing lummox to change how filters and/or layering work. In the meantime it seems that items are all slightly translucent.

<details>
<summary>Pictures (Updated!):</summary>

Dayshift:
![Neon-Carpet-Dayshift-2](https://user-images.githubusercontent.com/53582584/103258853-c3b92e00-494b-11eb-8120-e5f3e47ae09f.png)

Nightshift:
![Neon-Carpet-Nightshift-2](https://user-images.githubusercontent.com/53582584/103258855-c7e54b80-494b-11eb-8801-0ef6e6b73273.png)

In the dark:
![Neon-Carpet-Gridcheck-2](https://user-images.githubusercontent.com/53582584/103258859-cddb2c80-494b-11eb-93a8-384647a9202b.png)

</details>

<details>
<summary>Recipes:</summary>

All variants of simple neon carpets can be made with carpet, tinea luxor, and a dye:
- White: Made with sodium
- Red: Made with mindbreaker toxin
- Orange: Made with vitrium froth
- Yellow: Made with stabilizing agent
- Lime: Made with lime juice
- Green: Made with unstable mutagen
- Cyan: Made with salbutamol
- Teal: Made with nicotine
- Blue: Made with happiness
- Purple: Made with hyper-plasmium oxide
- Violet: Made with helbital
- Pink: Made with impedrezine
- Black: Made with multiver

</details>

<details>
<summary>TODOs:</summary>

- [X] TODO: Make the emissive decals on the neon carpet stacks dispay. The tile's emissive blocker appears to be blocking it.
  - Done: Neon carpet tiles have emissive decals. They tend to glow through other items but I can't fix that without lummox messing with how filters work.
- [X] TODO: Fix neon carpet emissive highlights showing up through structures. (Done, I had to rework emissive blockers a bit but it works!)
- [X] TODO: Fix neon carpet tile stacking. (Turns out I broke this a while ago.... whoops. Anyway, it's fixed now.)
- [X] TODO: Add neon carpet to the cargo imports. (The crate costs 3000 credits)
</details>

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- More aesthetic variety for people to decorate the station with.
- Gives coders the ability to define decal planes.
- Gives coders more layering options for emissive / emissive blocking objects.
  - Reworks vis_overlay updates a bit to make some structures (vending machines, computers, stasis beds) stop axing their own emissive blockers.
- Fixes a minor bug involving how stacks combine involving materials.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Neon carpets exists. Bug your local chemist or quartermaster for a supply.
code: Decals support user-defined planes and some forms of smoothing.
code: Emissive layers now support emissive turfs, items, and mobs as well as emissive blocking structures.
fix: You can recombine stacks of carpet you split apart again. Did anyone notice that this was broken?
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Adds emissive carpet, adds plane support to decal element

- Adds a basic abstract/debugging emissive carpet
- Makes decals capable of supporting plane
- Adds auto-smoothing decals

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

TemporalOroboros committed:
Adds simple neon carpet variations

TemporalOroboros committed:
Adds neon carpet reagents and recipes

Krysonism commented:
screenshot please

TemporalOroboros commented:
@Krysonism I have several. They're in the drop-downs indicated by the rightward facing triangles.

<details>
<summary>Like this:</summary>

See?
</details>

RaveRadbury commented:
Do these carpets damage nightmares, and if so can they be hit with the lighteater?

TemporalOroboros commented:
@RaveRadbury The carpets do not damage nightmares. At the moment they're just for decoration.

optimumtact commented:
now someone just has to add Mrs Pacman Generator game for ghosts to centcom z level

TemporalOroboros committed:
Refactors emissive blockers to allow multiple layers of emissive / emissive blocking objects

- Splits the emissive and emissive blocker plane masters into several plane masters which handle different layers of emissiveness.

TemporalOroboros committed:
Makes neon carpet tile stacks emissive

TemporalOroboros committed:
Rearranges and docs some emissive plane masters

- Folds the overlay lighting plane master into the emissive planes since it is also used to mask the lighting plane

TemporalOroboros committed:
Merge branch 'master' into upstream/master

TemporalOroboros committed:
Fixes null mats_per_unit stack recombining after splitting

- I think I broke this a while ago when I reworked how stacks handle materials. Whoops.
- This basically only effects carpet at the moment. Good thing I did this when I did!

TemporalOroboros committed:
Adds neon carpets to cargo

- Adds a cargo supply crate containing a _lot_ of neon carpets for 3000 credits

TemporalOroboros committed:
Fixes neon carpet highlights leaking through vending machines and such

- Turns out vending machines axed their own emissive blockers whenever they updated their icon because they cleared their managed_vis_overlays...
- Generic emissive blocking has been elementized and some update_overlays procs have been straightened out.

TemporalOroboros committed:
Fixes id_arg_index for the emissive blocker element

Twaticus commented:
i think this could benefit from a radial menu. instead of having a bunch of recipes for different colors you could have one slightly larger recipe with the option to change to any color using radials.

TemporalOroboros commented:
@Twaticus I'm more interested in keeping it inline with the other carpets for now.

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

Twaticus commented:
the other carpets should do it too ngl

TemporalOroboros commented:
@Twaticus I'll be honest, it's mostly because I like the idea of people going on small quests for \<rare reagent\> to produce a carpet factory. Also, I want to save colorful reagent for a carpet that shifts through the color spectrum.

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

Tlaltecuhtli commented:
i dont think anyone enjoys the <rare reagent> quest to make blue carpet, as chemist have more interesting thing to automate and botany can make fairy grass which also produces light. i barely see the other carpets used outside of me buying the crate myself or mappers adding new carpets as 'improvement'

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

Rohesie commented on code/__DEFINES/layers.dm:
```diff
@@ -91,66 +91,88 @@
-
-#define LIGHTING_PLANE 15
-#define LIGHTING_LAYER 15
+#define EMISSIVE_BLOCKER_PLANE						10
+#define EMISSIVE_BLOCKER_RENDER_TARGET				"*EMISSIVE_BLOCKER_PLANE"
+#define TURF_EMISSIVE_BLOCKER_PLANE					11
+#define TURF_EMISSIVE_BLOCKER_RENDER_TARGET			"*EMISSIVE_BLOCKER_PLANE_TURF"
+#define STRUCTURE_EMISSIVE_BLOCKER_PLANE			12
+#define STRUCTURE_EMISSIVE_BLOCKER_RENDER_TARGET	"*EMISSIVE_BLOCKER_PLANE_STRUCTURE"
+#define ITEM_EMISSIVE_BLOCKER_PLANE					13
+#define ITEM_EMISSIVE_BLOCKER_RENDER_TARGET			"*EMISSIVE_BLOCKER_PLANE_ITEM"
+#define MOB_EMISSIVE_BLOCKER_PLANE					14
+#define MOB_EMISSIVE_BLOCKER_RENDER_TARGET			"*EMISSIVE_BLOCKER_PLANE_MOB"
+
+#define EMISSIVE_PLANE						20
+#define EMISSIVE_RENDER_TARGET				"*EMISSIVE_PLANE"
+#define EMISSIVE_TURF_PLANE					21
+#define EMISSIVE_TURF_RENDER_TARGET			"*EMISSIVE_PLANE_TURF"
+#define EMISSIVE_STRUCTURE_PLANE			22
+#define EMISSIVE_STRUCTURE_RENDER_TARGET	"*EMISSIVE_PLANE_STRUCTURE"
+#define EMISSIVE_ITEM_PLANE					23
+#define EMISSIVE_ITEM_RENDER_TARGET			"*EMISSIVE_PLANE_ITEM"
+#define EMISSIVE_MOB_PLANE					24
+#define EMISSIVE_MOB_RENDER_TARGET			"*EMISSIVE_PLANE_MOB"
+#define EMISSIVE_UNBLOCKABLE_PLANE			31
+#define EMISSIVE_UNBLOCKABLE_RENDER_TARGET	"*EMISSIVE_UNBLOCKABLE_PLANE"
```
Could you remove the extra tabs/spaces? They don't really add much of value.

TemporalOroboros replied:
Sure

TemporalOroboros replied:
Done

Rohesie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -1703,6 +1703,104 @@
+/datum/reagent/carpet/neon
+	name = "Neon Carpet"
+	description = "For those who like the 1980s, vegas, and debugging."
+	color = "#000000"
```
```suggestion
	color = COLOR_BLACK
```

TemporalOroboros replied:
Done

Rohesie commented on code/datums/elements/emissive_blocker.dm:
```diff
@@ -0,0 +1,33 @@
+	update_blocker(target, null)
+	return ..()
+
+/datum/element/emissive_blocker/Detach(datum/source, force)
+	. = ..()
+	UnregisterSignal(source, COMSIG_ATOM_UPDATE_OVERLAYS)
+
```
```suggestion
/datum/element/emissive_blocker/Detach(datum/source, force)
	UnregisterSignal(source, COMSIG_ATOM_UPDATE_OVERLAYS)
	return ..()

```
Better call parent after on unwrapper procs.

TemporalOroboros replied:
Done

Rohesie commented on code/_onclick/hud/plane_master.dm:
```diff
@@ -110,8 +156,43 @@
+	var/blocker_target = EMISSIVE_BLOCKER_RENDER_TARGET
+
+/// The emissive blocker plane master used by structures to block... something. Presumably whatever's under turfs. (NOTE: Not actually implemented because adding this to every turf would be expensive and there's no reason to use this yet)
+/atom/movable/screen/plane_master/emissive_blocker/turf
```
```suggestion
/atom/movable/screen/plane_master/emissive_blocker/turf_emissive
```
Avoid to repeat the `/turf` word for type paths

Rohesie commented on code/_onclick/hud/plane_master.dm:
```diff
@@ -110,8 +156,43 @@
+	blocker_target = ITEM_EMISSIVE_BLOCKER_RENDER_TARGET
+
+/// The emissive blocker plane master used by items to block emissive turfs, structures, items and overlays thereof
+/atom/movable/screen/plane_master/emissive_blocker/mob
```
```suggestion
/atom/movable/screen/plane_master/emissive_blocker/mob_emissive
```
same for `/mob`, and to be fair I'd do the same for `item` and `structure` as well.

TemporalOroboros replied:
Done

Rohesie commented:
@actioninja , could you take a peek at the emissive changes?

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

TemporalOroboros committed:
Commits @Rohsie's suggestions

actioninja reviewed (changes requested):
not a fan of adding emissive blocking overlays to everything + increasing the filter count on everything to 5. items only was somewhat of a compromise for performance reasons to begin with.

TemporalOroboros commented on code/game/machinery/firealarm.dm:
```diff
@@ -71,8 +71,6 @@
 
 /obj/machinery/firealarm/update_overlays()
 	. = ..()
-	SSvis_overlays.remove_vis_overlay(src, managed_vis_overlays)
```
I added it to the functionality of update_overlays.

actioninja replied:
I see, missed that part.

TemporalOroboros commented on code/datums/elements/decals/_decal.dm:
```diff
@@ -5,18 +5,18 @@
 
-/datum/element/decal/Attach(atom/target, _icon, _icon_state, _dir, _cleanable=FALSE, _color, _layer=TURF_LAYER, _description, _alpha=255)
+/datum/element/decal/Attach(atom/target, _icon, _icon_state, _dir, _cleanable=FALSE, _color, _layer=TURF_LAYER, _plane=FLOOR_PLANE, _description, _alpha=255)
 	. = ..()
-	if(!isatom(target) || !generate_appearance(_icon, _icon_state, _dir, _layer, _color, _alpha, target))
+	if(!isatom(target) || !generate_appearance(_icon, _icon_state, _dir, _layer, _plane, _color, _alpha, target))
```
You can absolutely have overlays with separate planes from the parent. If you couldn't the carpet wouldn't work and since it does I must conclude that you can.

actioninja replied:
I am suspect of this, and am going to leave this unresolved for now.
If it works as is, you're probably right, but I'd prefer to get some confirmation first. I was under the impression that part of the point of vis_overlays was to allow things on different planes.

optimumtact commented:
> 
> 
> not a fan of adding emissive blocking overlays to everything + increasing the filter count on everything to 5. items only was somewhat of a compromise for performance reasons to begin with.

what change are you requesting here

actioninja commented:
> 
> 
> > not a fan of adding emissive blocking overlays to everything + increasing the filter count on everything to 5. items only was somewhat of a compromise for performance reasons to begin with.
> 
> what change are you requesting here

So there's two main issues with this:

1. higher number of filters affects client performance in a very difficult to measure, quantify, and benchmark way. We don't know if plane master filters have any higher of a cost or not.
2. extra object instance per unique object + extra list entry for every instance is not great for memory.

For blocking the emissives on turfs, maybe the "game plane" could be used as a mask source.

Not sure if anything else is needed considering that unblockable was pretty much specifically intended for items and mobs.

optimumtact commented:
So you're denying the merge? I don't understand what you are trying to communicate

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

TemporalOroboros committed:
Merge branch 'master' into neon-carpet

LemonInTheDark commented:
@actioninja Do you have more to say on this or should it get a review

actioninja commented:
it's probably fine after some further though. Watch for people complaining about performance though.

TemporalOroboros commented:
Reading through this is painful in hindsight.

LemonInTheDark commented:
I hate everything I've made more then 5 months ago
